DeeperFunc/Router.swift

```diff
@@ -0,0 +1,206 @@
+//
+//  Router.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 26/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+public protocol Route: Equatable {
+    func deconstruct<A>(_ constructor: ((A) -> Self)) -> A?
+}
+
+public class Router<U: Route>: DeepLinkRouter, CustomStringConvertible {
+    public let scheme: String
+    public let rootDeepLinkHandler: AnyDeepLinkHandler<U>?
+    
+    private var route: RoutePattern<U, Path>?
+
+    public init(scheme: String, rootDeepLinkHandler: AnyDeepLinkHandler<Intent>) {
+        self.scheme = scheme
+        self.rootDeepLinkHandler = rootDeepLinkHandler
+    }
+
+    public var description: String {
+        return "\(type(of: self)):\n[\(route?.template ?? "")]"
+    }
+    
+    public func openURL(_ url: URL) -> U? {
+        guard
+            let components = url.routeComponents,
+            url.scheme == scheme,
+            let result = route?.parse(components),
+            result.rest.path.isEmpty
+            else { return nil }
+        
+        return result.match
+    }
+
+    private func add(_ route: RoutePattern<U, Path>) -> Router {
+        self.route = self.route.map({ oldValue in
```

Comment:
`self.route` wraps all individual routes. When matching or printing using this route is equivalent to iterating through array of routes.